src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -67,6 +67,351 @@ public static ControlFlowGraph Create(IBlockOperation body)
+                                outOfRangeBlocksToVisit: null, 
+                                continueDispatchAfterFilterOrFinally,
+                                dispatchedExceptionsFromRegions,
+                                out _);
+            continueDispatchAfterFilterOrFinally.Free();
+            dispatchedExceptionsFromRegions.Free();
+        }
+
+        private static BitVector MarkReachableBlocks(
+            ArrayBuilder<BasicBlock> blocks, 
+            int firstBlockOrdinal, 
+            int lastBlockOrdinal, 
+            ArrayBuilder<BasicBlock> outOfRangeBlocksToVisit, 
+            PooledDictionary<ControlFlowGraph.Region, bool> continueDispatchAfterFilterOrFinally,
+            PooledHashSet<ControlFlowGraph.Region> dispatchedExceptionsFromRegions,
+            out bool fellThrough)
+        {
+            var visited = BitVector.Empty;
+            var toVisit = ArrayBuilder<BasicBlock>.GetInstance();
+
+            fellThrough = false;
+            toVisit.Push(blocks[firstBlockOrdinal]);
+
+            do
+            {
+                BasicBlock current = toVisit.Pop();
+
+                if (current.Ordinal < firstBlockOrdinal || current.Ordinal > lastBlockOrdinal)
+                {
+                    outOfRangeBlocksToVisit.Push(current);
+                    continue;
+                }
+
+                if (visited[current.Ordinal])
+                {
+                    continue;
+                }
+
+                visited[current.Ordinal] = true;
+                current.IsReachable = true;
+                bool canThrow = false;
+                bool fallThrough = true;
+
+                foreach (IOperation operation in current.Statements)
+                {
+                    if (operationCanThrow(operation))
+                    {
+                        canThrow = true;
+                        break;
+                    }
+                }
+
+                (IOperation Condition, bool JumpIfTrue, BasicBlock.Branch Branch) conditional = current.Conditional;
+                if (conditional.Condition != null)
+                {
+                    if (conditional.Condition.ConstantValue.HasValue && conditional.Condition.ConstantValue.Value is bool constant)
+                    {
+                        if (constant == conditional.JumpIfTrue)
+                        {
+                            followBranch(current, conditional.Branch, ref canThrow);
+                            fallThrough = false;
+                        }
+                    }
+                    else
+                    {
+                        if (operationCanThrow(conditional.Condition))
+                        {
+                            canThrow = true;
+                        }
+
+                        followBranch(current, conditional.Branch, ref canThrow);
+                    }
+                }
+
+                if (fallThrough)
```

Comment:
>fallThrough [](start = 20, length = 11)

If the basic block contains a `throw` statement, how does `fallThrough` get set to `false`?

Reply:
Is the `throw` treated as the end of (transfer from) a basic block?

---
In reply to: [177499039](https://github.com/dotnet/roslyn/pull/25698#discussion_r177499039) [](ancestors = 177499039)

Reply:
> If the basic block contains a throw statement, how does fallThrough get set to false?

You are correct, there is some inaccuracy here, however it is not observable because fellThrough only makes a difference when the branch is StructuredExceptionHandling. Nerveless, I'll make an adjustment in the next PR

---
In reply to: [177499039](https://github.com/dotnet/roslyn/pull/25698#discussion_r177499039) [](ancestors = 177499039)

Reply:
> Is the throw treated as the end of (transfer from) a basic block?

Yes, throw terminates a block

---
In reply to: [177500974](https://github.com/dotnet/roslyn/pull/25698#discussion_r177500974) [](ancestors = 177500974,177499039)

Reply:
ExceptionDispatch_42 test demonstrates that despite the inaccuracy for the case when fall through branch is a throw branch, the behavior is still correct.

---
In reply to: [177501287](https://github.com/dotnet/roslyn/pull/25698#discussion_r177501287) [](ancestors = 177501287,177500974,177499039)